### PR TITLE
Convert jujulib to use bakery methods directly

### PR DESCRIPTION
**Pull request by hatched**

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA
Use the GUI to perform actions modified in jujulib. They should work as expected

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7273/

File: jujugui/static/gui/src/app/jujulib/charmstore.js [hunk 1]
```diff
@@ -447,18 +439,12 @@ var module = module;
...
-        // turn off redirect, we want to silently fail
-        // if the macaroon is valid.
-        false
-      );
+      const url = this._generatePath('whoami');
+      const headers = null;
+      const wrappedCallback =
+        jujulib._wrap(
+          jujulib._transformAuthObject.bind(this,callback), {parseJSON: true});
+      return this.bakery.get(url, headers, wrappedCallback);
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/charmstore.js]: This call previously used to "turn off redirect".So I guess we should now use `this.bakery.withoutDischarge().get(...` here.

File: jujugui/static/gui/src/app/jujulib/index.js [hunk 2]
```diff
@@ -25,97 +20,59 @@ var module = module;
...
   */
-  const _makeRequest = function(
-    bakery, url, method, params, callback, parse, redirect) {
-    const wrappedCallback = (err, response) => {
+  const serializeObject = function(obj) {
+    return Object.keys(obj).map(p =>
+        `${encodeURIComponent(p)}=${encodeURIComponent(obj[p])}`).join('&');
+  };
+  /**
+    @param {Function} callback The callback to wrap.
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/index.js]: Please specify the signature of the original callback.

File: jujugui/static/gui/src/app/jujulib/index.js [hunk 3]
```diff
@@ -25,97 +20,59 @@ var module = module;
...
+      if (!options.parseJSON) {
         callback(null, text);
         return;
       }
-      let jsonResponse;
+      let jResp;
       try {
-        jsonResponse = JSON.parse(text);
+        jResp = JSON.parse(text);
+        err = jResp.error || jResp.Error || jResp.message || jResp.Message;
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/index.js]: Why are we considering a response with a message as an error? In pre-existing code we only lookup Error and error.

File: jujugui/static/gui/src/app/jujulib/payment.js [hunk 4]
```diff
@@ -582,11 +614,26 @@ var module = module;
...
+        try {
+          resp = JSON.parse(response);
+          const parsedError =
+            resp.error || resp.Error || resp.message || resp.Message;
+          if (parsedError) {
+            callback(parsedError, null);
+            return;
+          }
+          callback(null, resp);
+        } catch (e) {
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/payment.js]: s/e/_/

File: jujugui/static/gui/src/app/jujulib/terms.js [hunk 5]
```diff
@@ -113,29 +115,31 @@ var module = module;
...
+            termname: term.name,
+            termrevision: term.revision
+          };
+          if (term.owner) {
+            args.termowner = term.owner;
+          }
+          return args;
+        }));
+      const headers = {
+        'Content-Type': 'application/json'
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/terms.js]: `'Content-Type': jujulib.JSON_CONTENT_TYPE`? [shrug]

File: jujugui/static/gui/src/app/jujulib/test-index.js [hunk 6]
```diff
@@ -5,38 +5,56 @@
...
+    });
+
+    it(of + 'when json parsing fails', () => {
+      const stub = sinon.stub();
+      wrap(stub, {parseJSON: true})(
+        null, {target: {responseText: '{"invalid "json}'}});
+      assert.equal(
+        stub.args[0][0],
+        'SyntaxError: Unexpected token j in JSON at position 11');
+      assert.equal(stub.args[0][1], null);
```
Comment (frankban) [inline jujugui/static/gui/src/app/jujulib/test-index.js]: Nice tests!

File: jujugui/static/gui/src/app/jujulib/index.js (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/index.js]: Related to my email about the various service response values...I really just added it to maintain consistency. I can remove it and update the tests but I felt it best not to lose the functionality until we know what the services return for sure.

https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/jujulib/index.js#L82

File: jujugui/static/gui/src/app/jujulib/terms.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/terms.js]: We don't have any constants in jujulib yet, but it's a good idea, I'll make a note for a follow-up.

File: jujugui/static/gui/src/app/jujulib/payment.js (hunk 4 above)
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/payment.js]: done

File: jujugui/static/gui/src/app/jujulib/index.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/index.js]: done

File: jujugui/static/gui/src/app/jujulib/terms.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/terms.js]: Just to clarify - we have constants on a per file basis, but nothing global to jujulib.

File: jujugui/static/gui/src/app/jujulib/charmstore.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/charmstore.js]: updated

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7275/

**Comment by hatched** [comment]
Merging so that I can continue on with the remaining failing tests. Any extra comments can be addressed as a follow-up.

**Pull request merged**